Add Support for JetBrains Runtime

gmitch215 opened the pull request:
**Current Workaround**
```yml
- name: Setup JBR 21
  uses: gmitch215/setup-java@4e56c31b28053b7081e102121153633227d6dd8e
  with:
    distribution: 'jetbrains'
    java-version: 21
    cache: 'gradle'
  # For GitHub API (use if you believe you'd run into rate limits)
  env:
    GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
````

**Description:**
This PR adds support for the [JetBrains Runtime](https://github.com/JetBrains/JetBrainsRuntime), which is required in dependencies like [jewel](https://github.com/JetBrains/jewel) and useful for [Compose Multiplatform](https://www.jetbrains.com/lp/compose-multiplatform/).

The PR uses the binaries listed in `Binaries for Developers` , using the `JBRSDK` flavor as shown in the [releases page](https://github.com/JetBrains/JetBrainsRuntime/releases) for the default flavor `jdk`. Alternative flavors include `jdk+jcef` and `jdk+ft` with `jre` counterparts. 

It uses the GitHub API to find and parse versions, then calculates the URL using `cache-redirector.jetbrains.com` with the parsed SemVer, OS, Arch, and build number, which all output a `.tar.gz` file. Left a few comments in there for maintainability. Some tags have inconsistent naming conventions, or have been removed from public view for various reasons, so it will filter out those versions.

Since this uses the GitHub API, you can optionally specify a `GITHUB_TOKEN` environmental variable to increase your rate limit.

**Related issue:**
- Closes #399
- Closes #585 (Duplicate)

**Check list:**
- [x] Mark if documentation changes are required.
- [x] Mark if tests were added or updated to cover the changes.

**Notes**
Tested using `jetbrains-installer.test.ts`. Most of the tests were copied from other installers, but some I couldn't figure out how to bring over, so additional work may be required. I also copied `downloadTool` from the Temurin installer.

I also executed `npm run build` for `dist/setup/index.js`. I think something else was also thrown because I was on TypeScript 5.3, so I'm not sure how that will affect `ncc`.

Goooler commented:
I'm using your branch as a workaround

```yml
  # TODO: replace this once https://github.com/actions/setup-java/pull/637 gets merged.
  - uses: gmitch215/setup-java@07ea605e4b1f0f8b0621155fb1722d6094c3e183
    with:
      distribution: 'jetbrains'
      java-version: 17
```

with an error:

```sh
Installed distributions
  Error: No supported distribution was found for input jetbrains
```

is there anything I can do?

See:
- https://github.com/Goooler/kotlin-explorer/actions/runs/9574788494/job/26398609733
- https://github.com/Goooler/kotlin-explorer/commit/abfa1e3d127d3c8c3edfbf5c57d70068845f9773

gmitch215 commented:
Hmm, I'll debug and get back here later.

gmitch215 commented:
Seems I just forgot to add it to the map of recognized runtimes to their IDs. Update your commit hash and it should succeed.

My project passed with the following configuration:
```yml
      - name: Setup JBR 17
        uses: gmitch215/setup-java@7f3ab2648f3d20faa20c4fe6e52a455757e0f38a
        with:
          distribution: 'jetbrains'
          java-version: 17
          cache: 'gradle'
```
Output:
![image](https://github.com/actions/setup-java/assets/54124162/0a080ebe-fcff-4ff5-8525-f52e2513783d)

Goooler commented:
It could be picked correctly now!

But seems some libs are missing in this JDK installed.

```sh
Execution failed for task ':checkRuntime'.
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
> Failed to check JDK distribution: 'jlink', 'jpackage' are missing

  JDK distribution path: /Users/runner/hostedtoolcache/Java_JetBrains_jdk/17.0.11-1312.2/arm64/Contents/Home
```

See https://github.com/Goooler/kotlin-explorer/actions/runs/9575738587/job/26401058613.

Goooler commented:
The JDK installed by Gradle `toolchain` worked fine.

```
Invalid Java installation found at '/Users/runner/.gradle/jdks/jetbrains_s_r_o_-17-aarch64-os_x/jbrsdk_jcef-17.0.11-osx-aarch64-b1207.24/Contents/Home' (Auto-provisioned by Gradle). It will be re-checked in the next build. This might have performance impact if it keeps failing. Run the 'javaToolchains' task for more details.
```

Seems we need `Binaries for developers` instead of `Binaries for launching IntelliJ IDEA`, which is installed by your action. Check https://github.com/JetBrains/JetBrainsRuntime/releases/tag/jbr-release-17.0.11b1312.2.

gmitch215 commented:
I switched the URL to the ones listed in the `Binaries for Developers` section, still getting errors. Going to do some more research.

gmitch215 commented:
I guessed correctly! `:checkRuntime` seems to pass.

```yml
      - name: Setup JBR 17
        uses: gmitch215/setup-java@99fc2135f7f7b08068e180cb5f29340f9de70720
        with:
          distribution: 'jetbrains'
          java-version: 17
          cache: 'gradle'
```

(This action failed for unrelated reasons)

![image](https://github.com/actions/setup-java/assets/54124162/ce8e9b4b-644b-46b3-9670-656c23276b0f)

Goooler commented:
99fc2135f7f7b08068e180cb5f29340f9de70720 works for me now, someone who needs this can ref https://github.com/romainguy/kotlin-explorer/pull/62.

Thanks for your quick response!

gmitch215 commented:
Got `windows-latest` to build now. Everything else works fine too.

```yml
      - name: Setup JBR 17
        uses: gmitch215/setup-java@e1f4c15facc0cf68c0e764340e374be034e4a859
        with:
          distribution: 'jetbrains'
          java-version: 17
          cache: 'gradle'
```

eskatos commented:
Thank you @gmitch215, I'm using your branch as a workaround too and can confirm it works fine!
It would be great if this contribution could be accepted.

eskatos commented:
On a relatively complex workflow with a fan-out setup and composite actions that parallelize builds I get rate limit failures from time to time:
```
Run gmitch215/setup-java@36531813787dd0ff3269feeb1144b35b3d68f927
  with:
    java-version: 21
    java-package: jdk
    distribution: jetbrains
    check-latest: false
    server-id: github
    server-username: GITHUB_ACTOR
    server-password: GITHUB_TOKEN
    overwrite-settings: true
    job-status: success
    token: ***
Installed distributions
  Trying to resolve the latest version from remote
  Error: API rate limit exceeded for 52.159.14[7](https://github.com/eskatos/private-repository/actions/runs/1234567890/job/1234567890#step:3:8).63. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)
```

Is there something that can be done in this action's implementation or on my side?

gmitch215 commented:
@eskatos The JetBrains `cache-redirector` server does rate limit by IP. To my knowledge, there isn't a way to change the rate limit. The root URL's info page is internal as well, and only accessible by employees.

Your best bet is to do something that changes the IP (not sure if re-running it manually does this, but worth a shot). Unless someone knows how the runner could provide authentication.

eskatos commented:
Re-running did make the job pass indeed.

> This error page is about the GitHub API, which is something I need to add.

Thank you for fixing this!

eskatos commented:
With b4d76a354bef7187244f74eea4e80e674a7b4b55 I'm getting this:
```
  Trying to resolve the latest version from remote
  Error: Bad credentials
```

gmitch215 commented:
I forgot GitHub doesn't automatically inject the `GITHUB_TOKEN` environment variable on action workflows.

The action should now check if it is provided. Update to this:
```yml
- name: Setup JBR 21
  uses: gmitch215/setup-java@6d2c5e1f82f180ae79f799f0ed6e3e5efb4e664d
  with:
    distribution: 'jetbrains'
    java-version: 21
    cache: 'gradle'
  env:
    GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
````

eskatos commented:
Thanks, works for me!

gmitch215 commented:
+1. Will try to resolve the merge conflicts at some point.

HarithaVattikuti commented:
Hello @gmitch215 , We are working to prioritize this feature. We will update you once we have some feedback.

Jonatha1983 commented:
Hi, @gmitch215  - thanks for your work!!!
Small question: for now, we should keep using 
```
uses: gmitch215/setup-java@6d2c5e1f82f180ae79f799f0ed6e3e5efb4e664d
```
Till this is merged? 

Thanks again

gmitch215 commented:
> Hi, @gmitch215  - thanks for your work!!!
> 
> Small question: for now, we should keep using 
> 
> ```
> 
> uses: gmitch215/setup-java@6d2c5e1f82f180ae79f799f0ed6e3e5efb4e664d
> 
> ```
> 
> Till this is merged? 
> 
> 
> 
> Thanks again 

Sure. The most recent commit for the workaround is available in the top message of this PR.

gmitch215 commented:
@HarithaVattikuti been 2 weeks since we last spoke, any updates?

mahabaleshwars commented:
Hi @gmitch215, a few checks are failing in the PR:

Build the project using Prettier: Ensure that you run Prettier to format the code properly.

JDK 11 Naming Convention Issue: When using JDK 11, the system expects the version to be in the format 11_0_16b2043.64 instead of 11.0.16-b2043.64. This discrepancy is causing the current failure.

Please update the PR once these issues are resolved.

gmitch215 commented:
@mahabaleshwars Issues fixed. I also added a test suite that checks if the URLs return 403 (access denied, meaning not found) instead of 200/302.

mahabaleshwars commented:
Hello @gmitch215,

Many of the checks are still failing, including those for JDK 11. Additionally, some links requested for JDK 11 are not available. For example, when tried on a Windows runner, the following URL was requested from the workflow: https://cache-redirector.jetbrains.com/intellij-jbr/jbrsdk_jcef-11.0.13-windows-x64-b1504.49.tar.gz, but it is not available at the source. Could you please rectify these issues and update the PR?

Before committing, please run the following commands in your project to ensure proper formatting and build success:

npm run format
npm run build
npm run test
These commands will handle Prettier formatting and verify the distribution (Dist).

Thank you!

gmitch215 commented:
Switched to plain `JBRSDK` and fixed the URL.

aparnajyothi-y commented:
Hello @gmitch215, Many of the checks are still failing, including those for JDK 11. Please review the changes on your end and address these issues. Once updated, kindly update the PR.

mahabaleshwars commented:
Hello @gmitch215,

Could you please confirm if the issues have been resolved and the PR has been updated? We would like to proceed with testing and approval once it's ready.

gmitch215 commented:
The issues haven't been resolved yet - I'm still working on trying to figure out what the issues are with JDK 11 specifically. I'll get back to you once I've completed it.

gmitch215 commented:
Try this build. I also added some more tests to `e2e-versions.yml` along with including the `GITHUB_TOKEN` to prevent rate limits.

gmitch215 commented:
It appears that sometimes JetBrains will make certain URLs inaccessible for some reason, so I check with `HEAD` requests to make sure they return a `200`. Running `npm run test` didn't show any significant speed differences.

gmitch215 commented:
This one should work. There is an additional `_nomod` suffix for some versions, and it now properly filters out invalid versions.

I also added some new package types:
- `jdk` for `jbrsdk`
- `jdk+jcef` for `jbrsdk with JCEF`
- `jdk+ft` for `jbrsdk (FreeType)`

Along with its JRE counterparts.

There are also some GitHub API-related issues in `e2e-versions.yml`. This should include `GITHUB_TOKEN` as an environmental variable.

gmitch215 commented:
Final issues were rate limits and issues `e2e-versions.yml`. Added `GITHUB_TOKEN` environmental variable to those issues.

The fail on `macos-latest` is a GraalVM issue and is not relevant to this PR.

gmitch215 commented:
@aparnajyothi-y @mahabaleshwars Any updates on getting this approved now that all the checks have passed?

mahabaleshwars commented:
Hi @gmitch215,

The updated PR resolves most issues, but there are some version-specific discrepancies as per this [documentation](https://github.com/JetBrains/JetBrainsRuntime/tree/jbr11_0_16b2043.64?tab=readme-ov-file#releases-based-on-jdk-11) :

Tests fail for certain versions (e.g., 11_0_15) but work for others (e.g., 11.0.15).
Users may input versions with underscores (e.g., 11_0_15) instead of dots (e.g., 11.0.15), which should be addressed.
Some versions (e.g., 11.0.10) do not work, while others (e.g., 11.0.11) do.
It would be beneficial to update the setup-java documentation to reflect these scenarios, especially the version formatting and compatibility issues.

gmitch215 commented:
> Hi @gmitch215,
> 
> The updated PR resolves most issues, but there are some version-specific discrepancies as per this [documentation](https://github.com/JetBrains/JetBrainsRuntime/tree/jbr11_0_16b2043.64?tab=readme-ov-file#releases-based-on-jdk-11) :
> 
> Tests fail for certain versions (e.g., 11_0_15) but work for others (e.g., 11.0.15). Users may input versions with underscores (e.g., 11_0_15) instead of dots (e.g., 11.0.15), which should be addressed. 

I am a little confused. Are you saying `11_0_15` doesn't work but `11.0.15` does? They're all based on the GitHub tags the bot publishes. I'll validate whether underscores are inputted, since that behavior isn't intended, but that behavior should be expected.

> Some versions (e.g., 11.0.10) do not work, while others (e.g., 11.0.11) do. It would be beneficial to update the setup-java documentation to reflect these scenarios, especially the version formatting and compatibility issues.

JetBrains seems to have removed `11.0.10` from their releases page, despite being available via URL. I'll have to add it manually.

aparnajyothi-y commented:
Hi @gmitch215, could you please confirm if you are working on updating the documentation to include information about the valid input for JDK 11?

gmitch215 commented:
> Hi @gmitch215, could you please confirm if you are working on updating the documentation to include information about the valid input for JDK 11?

Yes, currently working on new fixes.

crowforkotlin commented:
hello, I use jewel in my project, which requires jbrsdk+jcef but I see that jbr doesn't have jpackage, so when I use ./gradlew packageDmg, I get an error! Failed to check JDK distribution: 'jpackage' is missing
Is there a way around this?

gmitch215 commented:
> hello, I use jewel in my project, which requires jbrsdk+jcef but I see that jbr doesn't have jpackage, so when I use ./gradlew packageDmg, I get an error! Failed to check JDK distribution: 'jpackage' is missing Is there a way around this?

You can use `jbr+jcef` as the `distribution` parameter and it should install it. I'll update the documentation soon.

gmitch215 commented on .github/workflows/e2e-versions.yml:
```diff
@@ -125,7 +136,14 @@ jobs:
         os: [macos-latest, windows-latest, ubuntu-latest]
         distribution:
-          ['temurin', 'zulu', 'liberica', 'dragonwell', 'sapmachine']
+          [
+            'temurin',
+            'zulu',
+            'liberica',
+            'dragonwell',
+            'sapmachine',
+            'jetbrains'
+          ]
```
That's how the original file was.

jsoref replied:
the advantage of this notation (assuming it works) is that the next person to add an entry won't be blamed for changing your line (in the `[...]` notation, they'd need to add a `,` before they add they entry).

Some things seem to use the notation, so I don't think the project is opposed to it...

gmitch215 replied:
Right, but I don't want to change something that isn't apart of this PR.

gmitch215 commented:
@aparnajyothi-y @mahabaleshwars Documentation is updated and hidden versions were added.

crowforkotlin commented:
> > hello, I use jewel in my project, which requires jbrsdk+jcef but I see that jbr doesn't have jpackage, so when I use ./gradlew packageDmg, I get an error! Failed to check JDK distribution: 'jpackage' is missing Is there a way around this?您好，我在项目中使用了 jewel，这需要 jbrsdk+jcef，但我看到 jbr 没有 jpackage，所以当我使用 ./gradlew packageDmg 时，我收到一个错误！无法检查 JDK 发行版：缺少“jpackage”有没有办法解决这个问题？
> 
> You can use `jdk+jcef` as the `package-type` parameter and it should install it. I'll update the documentation soon.您可以使用 `jdk+jcef` 作为 `package-type` 参数，它应该会安装它。我很快就会更新文档。



> > hello, I use jewel in my project, which requires jbrsdk+jcef but I see that jbr doesn't have jpackage, so when I use ./gradlew packageDmg, I get an error! Failed to check JDK distribution: 'jpackage' is missing Is there a way around this?您好，我在项目中使用了 jewel，这需要 jbrsdk+jcef，但我看到 jbr 没有 jpackage，所以当我使用 ./gradlew packageDmg 时，我收到一个错误！无法检查 JDK 发行版：缺少“jpackage”有没有办法解决这个问题？
> 
> You can use `jdk+jcef` as the `package-type` parameter and it should install it. I'll update the documentation soon.您可以使用 `jdk+jcef` 作为 `package-type` 参数，它应该会安装它。我很快就会更新文档。

how to configure this ‘package-type’

crowforkotlin commented:
> > hello, I use jewel in my project, which requires jbrsdk+jcef but I see that jbr doesn't have jpackage, so when I use ./gradlew packageDmg, I get an error! Failed to check JDK distribution: 'jpackage' is missing Is there a way around this?
> 
> You can use `jdk+jcef` as the `package-type` parameter and it should install it. I'll update the documentation soon.

how to configure this ‘package-type’

crowforkotlin commented:
> @aparnajyothi-y @mahabaleshwars Documentation is updated and hidden versions were added.

Maybe I didn't make my question clear to you, you can look here https://github.com/JetBrains/jewel/issues/709

gmitch215 commented:
> > @aparnajyothi-y @mahabaleshwars Documentation is updated and hidden versions were added.
> 
> Maybe I didn't make my question clear to you, you can look here [JetBrains/jewel#709](https://github.com/JetBrains/jewel/issues/709)

```yml
- name: Setup JBR 21
  uses: gmitch215/setup-java@main
  with:
    distribution: 'jetbrains'
    java-version: 21
    cache: 'gradle'
    package-type: 'jdk+jcef'
  # For GitHub API (use if you believe you'd run into rate limits)
  env:
    GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

crowforkotlin commented:
> > > @aparnajyothi-y @mahabaleshwars Documentation is updated and hidden versions were added.
> > 
> > 
> > Maybe I didn't make my question clear to you, you can look here [JetBrains/jewel#709](https://github.com/JetBrains/jewel/issues/709)
> 
> ```yaml
> - name: Setup JBR 21
>   uses: gmitch215/setup-java@main
>   with:
>     distribution: 'jetbrains'
>     java-version: 21
>     cache: 'gradle'
>     package-type: 'jdk+jcef'
>   # For GitHub API (use if you believe you'd run into rate limits)
>   env:
>     GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
> ```

ok, thank you

Jonatha1983 commented:
@gmitch215  its package-type or java-package? 
I am getting this: 

Unexpected input(s) 'package-type', valid inputs are ['java-version', 'java-version-file', 'distribution', 'java-package', 'architecture', 'jdkFile', 'check-latest', 'server-id', 'server-username', 'server-password', 'settings-path', 'overwrite-settings', 'gpg-private-key', 'gpg-passphrase', 'cache', 'cache-dependency-path', 'job-status', 'token', 'mvn-toolchain-id', 'mvn-toolchain-vendor']

gmitch215 commented:
> @gmitch215 its package-type or java-package? I am getting this:
> 
> Unexpected input(s) 'package-type', valid inputs are ['java-version', 'java-version-file', 'distribution', 'java-package', 'architecture', 'jdkFile', 'check-latest', 'server-id', 'server-username', 'server-password', 'settings-path', 'overwrite-settings', 'gpg-private-key', 'gpg-passphrase', 'cache', 'cache-dependency-path', 'job-status', 'token', 'mvn-toolchain-id', 'mvn-toolchain-vendor']

That's correct, my bad.

mahabaleshwars commented:
Hello @gmitch215,

Thanks again for the continuous improvements to the PR! The latest changes addressing the issue with JDK 11.0.10 and the documentation look great. However, I’ve noticed a couple of things that still need attention:

There are some CI failures that can be fixed by running the command:
```npm audit fix```

Regarding the JDK version input, the documentation mentions that versions are based on the GitHub tags on the JetBrains Runtime releases page, with both underscores (11_0_13) and dots (11.0.13) supported. While the dot format is encouraged for consistency, both are valid. That said, when the input for JDK 11 contains underscores, it is still failing. I see that you've already addressed this, but could you please take another look to see if there’s something we might have missed? I’ve attached a screenshot for reference.

<img width="1779" alt="Screenshot 2024-12-03 at 7 09 52 PM" src="https://github.com/user-attachments/assets/756d737d-b7a8-4974-a8f3-53096b79aefa">


Thanks again for your work on this!

gmitch215 commented:
@mahabaleshwars The difference between using `_` and `.` is usually inconsistent. Sometimes it works and sometimes it doesn't. 

That being said, I notice some tags *do* exist with underscore notation (`_`), so I'll take a look and update code/documentation. 

Thank you all for moving this forward!

gmitch215 commented:
@mahabaleshwars 

Dot notation is enforced by the base installer, so underscore support is not possible. Updated the documentation to reflect this and fixed a typo.

Also ran `npm audit fix`.

gmitch215 commented:
Once again, `macos-latest` issues are with GraalVM.

mahabaleshwars commented:
Hello @gmitch215, CI workflow issues are fixed now and this should address the issues in this PR. Please sync the main branch with setup-java.

gmitch215 commented:
@mahabaleshwars Done.

priyagupta108 commented:
Hi @gmitch215 👋,

For most Java distributions (such as Temurin, GraalVM, Zulu, SapMachine, etc.), pre-release (EA) versions are not installed unless explicitly requested. However, we’ve observed that with `JetBrainsRuntime`, if a pre-release version is available, it gets installed by default even when only a standard version like 21 is specified. Below is a screenshot of the run showing this behavior.

<img width="1078" alt="Screenshot 2025-05-23 at 2 38 46 PM" src="https://github.com/user-attachments/assets/1cb3d8f7-f9c4-429f-8b17-d56a7fb4a9c3" /><br>


Could you confirm if this is the intended behavior? Ideally, `JetBrainsRuntime` should follow the same convention as other distributions and install only stable releases unless an EA version is specifically requested.
If this is not intended, we plan to update our process to always filter out pre-release versions unless explicitly requested.

Looking forward to your response.

gmitch215 commented:
Hello @priyagupta108,

Thanks for letting me know. I was unaware that this was the intended behavior. I can maybe submit a PR filtering out pre-releases if not requested, but I'm a bit busy at the moment, so I would need to do it later.

priyagupta108 commented:
Hello @gmitch215, 
Thank you for your quick response and for being open to helping out.
To keep things consistent with how other distributions are handled, we’ll move ahead with updating the process to filter out prereleases unless they’re specifically requested. If you have any suggestions or preferences regarding the implementation, please feel free to share your thoughts.

gmitch215 commented:
@priyagupta108 No problem, glad I could help. I know the GitHub API has a `prelease` value from its API, so it shouldn't be too much new code.
